modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/utils.py

```diff
@@ -250,18 +256,32 @@ def check_cols_to_join(what, df, col_names):
     Tuple[HdkOnNativeDataframe, list]
         The aligned data frame and column names.
     """
-    cols = df.columns
+
+    def check_dtype(name, dtype):
+        if not (
+            is_string_dtype(dtype)
+            or is_any_int_dtype(dtype)
+            or is_datetime64_dtype(dtype)
+            or is_categorical_dtype(dtype)
+        ):
+            raise NotImplementedError(f"Join on column '{name}' of type '{dtype}'")
```

Comment:
shouldn't we somehow compare that dtypes for the merging keys are equal instead? As far as I understand HDK doesn't support merging on mixed dtypes, however pandas can sometimes perform the merge:
```python
# here I'm running on your branch from the pull request
>>> import modin.pandas as pd
>>> df1 = pd.DataFrame({"a": [1, 2, 3], "b": [3, 4, 5]})
>>> df2 = pd.DataFrame({"a": [3, 2, 4]}, dtype="category")
>>> df1._to_pandas().merge(df2._to_pandas())
   a  b
0  2  4
1  3  5
>>> df1.merge(df2) 
RuntimeError: Cannot compare between INT64 and DICT32(TEXT)[16777217] # this error comes from HDK
```

Reply:
Good catch!